docs/resources/server/advancements.md

```diff
@@ -43,41 +43,38 @@ A list of criteria triggers defined by vanilla can be found in `CriteriaTriggers
 
-Conditions are usually passed in through the constructor. The `AbstractCriterionTriggerInstance` super constructor requires the instance to define the registry name of the trigger and the conditions the player must meet as an `ContextAwarePredicate`. The registry name of the trigger should be supplied to the super directly while the conditions of the player should be a constructor parameter.
+Conditions are usually passed in through the constructor. The `SimpleCriterionTrigger.SimpleInstance` interface requires only one function, called `player()`, which returns the conditions the player must meet as an `Optional<ContextAwarePredicate>`. If the subclass is a Java record, the automatically generated `player()` method will suffice.
 
 ```java
-// Where ID is the registry name of the trigger
-public ExampleTriggerInstance(ContextAwarePredicate player, ItemPredicate item) {
-  super(ID, player);
-  // Store the item condition that must be met
+public record ExampleTriggerInstance(Optional<ContextAwarePredicate> player, ItemPredicate item) implements SimpleCriterionTrigger.SimpleInstance {
+  // extra methods here
 }
 ```
 
 :::note
-Typically, trigger instances have a static constructor which allow these instances to be easily created for data generation. These static factory methods can also be statically imported instead of the class itself.
+Typically, trigger instances have static helper methods which construct the full `Criterion<T>` object from the arguments to the instance. This allows these instances to be easily created during data generation, but are optional.
 
 ```java
-public static ExampleTriggerInstance instance(ContextAwarePredicate player, ItemPredicate item) {
-  return new ExampleTriggerInstance(player, item);
+public static Criterion<ExampleTriggerInstance> instance(ContextAwarePredicate player, ItemPredicate item) {
+  return EXAMPLE_TRIGGER.get().createCriterion(new ExampleTriggerInstance(Optional.of(player), item));
 }
 ```
 :::
 
-Additionally, the `#serializeToJson` method should be overridden. The method should add the conditions of the instance to the other JSON data.
+While it could be defined elsewhere, it is also typical to define a [`Codec`][codec] which can serialize this class as a constant in the class.
+A codec which can serialize this class must be defined somewhere, as it is required to implement `SimpleCriterionTrigger#codec` and is used to serialize criteria to JSON during data generation and deserialize them during loading.
 
 ```java
-@Override
-public JsonObject serializeToJson(SerializationContext context) {
-  JsonObject obj = super.serializeToJson(context);
-  // Write conditions to json
-  return obj;
-}
+public static final Codec<ExampleTriggerInstance> CODEC = RecordCodecBuilder.create(instance -> instance.group(
+  ContextAwarePredicate.CODEC.optionalFieldOf("player").forGetter(ExampleTriggerInstance::player),
```

Comment:
`Optional`s for trigger instances are usually wrapped via `ExtraCodecs#strictOptionalField` instead of using `#optionalFieldOf`. This is because the latter will always be mapped as a success, regardless of if parsing fails. Meanwhile, if the former fails to parse, then an error result will be provided instead.

Reply:
I didn't understand the distinction myself, thanks